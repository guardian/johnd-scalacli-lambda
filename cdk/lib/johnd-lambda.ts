import { GuApiLambda } from "@guardian/cdk";
import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import type { App } from "aws-cdk-lib";
import {Runtime} from "aws-cdk-lib/aws-lambda";
import type {CfnFunction} from "aws-cdk-lib/aws-sam";

export class JohndLambda extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps, buildNumber: string) {
    super(scope, id, props);

    const lambda = new GuApiLambda(this, "my-lambda", {
      fileName: `johnd-lambda-${buildNumber}.jar`,
      handler: "Handler::handleRequest",
      runtime: Runtime.JAVA_17,
      monitoringConfiguration: {
        noMonitoring: true,
      },
      app: "johnd-lambda",
      api: {
        id: "johnd-lambda-api",
        description: `This is an example AAAAPI ${buildNumber}`,
      },
      functionName: `johnd-lambda-${this.stage}`,
      enableVersioning: true,
    });
    (lambda.node.defaultChild as CfnFunction).addPropertyOverride("SnapStart", { ApplyOn: "PublishedVersions" });

  }

}
