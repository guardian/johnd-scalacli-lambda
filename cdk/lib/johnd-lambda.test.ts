import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { JohndLambda } from "./johnd-lambda";

describe("The JohndLambda stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new JohndLambda(app, "JohndLambda", { stack: "playground", stage: "TEST" });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
