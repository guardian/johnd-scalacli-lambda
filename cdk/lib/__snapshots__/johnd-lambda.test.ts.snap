// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The JohndLambda stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuApiLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "mylambdajohndlambdaapiEndpoint72382377": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "mylambdajohndlambdaapi14D1279D",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "mylambdajohndlambdaapiDeploymentStageprod417561F3",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "mylambdaE03B2284": {
      "DependsOn": [
        "mylambdaServiceRoleDefaultPolicyBDADF1CE",
        "mylambdaServiceRoleEF0A78B7",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "playground/TEST/johnd-lambda/johnd-lambda.jar",
        },
        "Environment": {
          "Variables": {
            "APP": "johnd-lambda",
            "STACK": "playground",
            "STAGE": "TEST",
          },
        },
        "Handler": "Handler::handleRequest",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "mylambdaServiceRoleEF0A78B7",
            "Arn",
          ],
        },
        "Runtime": "java17",
        "Tags": [
          {
            "Key": "App",
            "Value": "johnd-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/johnd-scalacli-lambda",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "mylambdaServiceRoleDefaultPolicyBDADF1CE": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/playground/TEST/johnd-lambda/johnd-lambda.jar",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/playground/johnd-lambda",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/playground/johnd-lambda/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "mylambdaServiceRoleDefaultPolicyBDADF1CE",
        "Roles": [
          {
            "Ref": "mylambdaServiceRoleEF0A78B7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "mylambdaServiceRoleEF0A78B7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "johnd-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/johnd-scalacli-lambda",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "mylambdajohndlambdaapi14D1279D": {
      "Properties": {
        "Description": "This is an example API",
        "Name": "playground-TEST-johnd-lambda-api",
        "Tags": [
          {
            "Key": "App",
            "Value": "johnd-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/johnd-scalacli-lambda",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "mylambdajohndlambdaapiANY531D7241": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "mylambdaE03B2284",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Fn::GetAtt": [
            "mylambdajohndlambdaapi14D1279D",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "mylambdajohndlambdaapi14D1279D",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "mylambdajohndlambdaapiANYApiPermissionJohndLambdamylambdajohndlambdaapiC945B4DCANY2459BAF3": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "mylambdaE03B2284",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "mylambdajohndlambdaapi14D1279D",
              },
              "/",
              {
                "Ref": "mylambdajohndlambdaapiDeploymentStageprod417561F3",
              },
              "/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "mylambdajohndlambdaapiANYApiPermissionTestJohndLambdamylambdajohndlambdaapiC945B4DCANY740F6C06": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "mylambdaE03B2284",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "mylambdajohndlambdaapi14D1279D",
              },
              "/test-invoke-stage/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "mylambdajohndlambdaapiAccount883159BC": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "mylambdajohndlambdaapi14D1279D",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "mylambdajohndlambdaapiCloudWatchRole2534763D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "mylambdajohndlambdaapiCloudWatchRole2534763D": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "johnd-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/johnd-scalacli-lambda",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "mylambdajohndlambdaapiDeployment7FEB2ACFdbd6058dddc2d011d752f72bcc339d12": {
      "DependsOn": [
        "mylambdajohndlambdaapiproxyANY96FE1B08",
        "mylambdajohndlambdaapiproxyAC91C370",
        "mylambdajohndlambdaapiANY531D7241",
      ],
      "Properties": {
        "Description": "This is an example API",
        "RestApiId": {
          "Ref": "mylambdajohndlambdaapi14D1279D",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "mylambdajohndlambdaapiDeploymentStageprod417561F3": {
      "DependsOn": [
        "mylambdajohndlambdaapiAccount883159BC",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "mylambdajohndlambdaapiDeployment7FEB2ACFdbd6058dddc2d011d752f72bcc339d12",
        },
        "RestApiId": {
          "Ref": "mylambdajohndlambdaapi14D1279D",
        },
        "StageName": "prod",
        "Tags": [
          {
            "Key": "App",
            "Value": "johnd-lambda",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/johnd-scalacli-lambda",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "mylambdajohndlambdaapiproxyAC91C370": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "mylambdajohndlambdaapi14D1279D",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": {
          "Ref": "mylambdajohndlambdaapi14D1279D",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "mylambdajohndlambdaapiproxyANY96FE1B08": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "mylambdaE03B2284",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "mylambdajohndlambdaapiproxyAC91C370",
        },
        "RestApiId": {
          "Ref": "mylambdajohndlambdaapi14D1279D",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "mylambdajohndlambdaapiproxyANYApiPermissionJohndLambdamylambdajohndlambdaapiC945B4DCANYproxy095DA3BD": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "mylambdaE03B2284",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "mylambdajohndlambdaapi14D1279D",
              },
              "/",
              {
                "Ref": "mylambdajohndlambdaapiDeploymentStageprod417561F3",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "mylambdajohndlambdaapiproxyANYApiPermissionTestJohndLambdamylambdajohndlambdaapiC945B4DCANYproxy6D32CB3F": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "mylambdaE03B2284",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "mylambdajohndlambdaapi14D1279D",
              },
              "/test-invoke-stage/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
